src/lists.py

```diff
@@ -78,15 +78,32 @@ def __init__(self, seq: Iterable[T] = ()):
         # after that link.
         for val in seq:
             insert_after(self.head.prev, val)
-
-    def __str__(self) -> str:
-        """Get string with the elements going in the next direction."""
-        elms: list[str] = []
+    def __eq__(self, other):
+        try:
+            for x, y in zip(
+                self.into_generator(), other.into_generator(),
+                strict= True
+                ):
+                if x != y:
+                    return False
+        except ValueError:
+            return False
+        return True
```

Comment:
This feels a bit over-engineered to me. You are invoking a lot of features just to check if the two lists are pair-wise equal. If you just run through them, comparing pairs, you get the same result. Should one be shorter than the other, they don't reach their respective head at the same time. So, something like this I find more to my taste:

```python
    def __eq__(self, other: DLList):
        # you could test the class here if you want to...
        x, y = self.head.next, other.head.next
        while x != self.head and y != other.head:
            if x.val != y.val:
                return False
            x, y = x.next, y.next
        # If they are equal, they are both pointing at
        # their head now. If not, one was longer than
        # the other
        return x == self.head and y == other.head
```